go-controller/pkg/ovn/master.go

```diff
@@ -184,43 +181,24 @@ func (oc *Controller) upgradeToSingleSwitchOVNTopology(existingNodeList *kapi.No
 func (oc *Controller) upgradeOVNTopology(existingNodes *kapi.NodeList) error {
 	// Find out the current OVN topology version. If "k8s-ovn-topo-version" key in external_ids column does not exist,
 	// it is prior to OVN topology versioning and therefore set version number to OvnCurrentTopologyVersion
-	ver := 0
-	stdout, stderr, err := util.RunOVNNbctl("--data=bare", "--no-headings", "--columns=name", "find", "logical_router",
-		fmt.Sprintf("name=%s", types.OVNClusterRouter))
+	ver, err := util.DetermineOVNTopoVersion()
 	if err != nil {
-		return fmt.Errorf("failed in retrieving %s to determine the current version of OVN logical topology: "+
-			"stderr: %q, error: %v", types.OVNClusterRouter, stderr, err)
-	}
-	if len(stdout) == 0 {
-		// no OVNClusterRouter exists, DB is empty, nothing to upgrade
-		return nil
-	}
-
-	stdout, stderr, err = util.RunOVNNbctl("--if-exists", "get", "logical_router", types.OVNClusterRouter,
-		"external_ids:k8s-ovn-topo-version")
-	if err != nil {
-		return fmt.Errorf("failed to determine the current version of OVN logical topology: stderr: %q, error: %v",
-			stderr, err)
-	} else if len(stdout) == 0 {
-		klog.Infof("No version string found. The OVN topology is before versioning is introduced. Upgrade needed")
-	} else {
-		v, err := strconv.Atoi(stdout)
-		if err != nil {
-			return fmt.Errorf("invalid OVN topology version string for the cluster: %s", stdout)
-		} else {
-			ver = v
-		}
+		return err
 	}
 
 	// If current DB version is greater than OvnSingleJoinSwitchTopoVersion, no need to upgrade to single switch topology
-	if ver < OvnSingleJoinSwitchTopoVersion {
+	if ver < types.OvnSingleJoinSwitchTopoVersion {
 		klog.Infof("Upgrading to Single Switch OVN Topology")
 		err = oc.upgradeToSingleSwitchOVNTopology(existingNodes)
 	}
-	if err == nil && ver < OvnNamespacedDenyPGTopoVersion {
+	if err == nil && ver < types.OvnNamespacedDenyPGTopoVersion {
 		klog.Infof("Upgrading to Namespace Deny PortGroup OVN Topology")
 		err = oc.upgradeToNamespacedDenyPGOVNTopology(existingNodes)
 	}
+	// If version is less than Host -> Service with OpenFlow, we need to remove and cleanup DGP
+	if err == nil && ver < types.OvnHostToSvcOFTopoVersion {
+		err = cleanupDGP(existingNodes)
```

Comment:
like we discussed on slack, this function should be called only for the `sharedGatewayMode` as local gateway mode still depends on `node_local_switch`

Reply:
done